server/gateway.go

```diff
@@ -2050,6 +2027,38 @@ func (c *client) sendMsgToGateways(acc *Account, msg, subject, reply []byte, qgr
 	}
 }
 
+func (s *Server) gatewayHandleServiceImport(acc *Account, subject []byte, change int32) {
+	sid := make([]byte, 0, len(acc.Name)+len(subject)+1)
+	sid = append(sid, acc.Name...)
+	sid = append(sid, ' ')
+	sid = append(sid, subject...)
+	sub := &subscription{subject: subject, sid: sid}
+
+	var rspa [1024]byte
+	rsproto := rspa[:0]
+	if change > 0 {
+		rsproto = append(rsproto, rSubBytes...)
+	} else {
+		rsproto = append(rsproto, rUnsubBytes...)
+	}
+	rsproto = append(rsproto, ' ')
+	rsproto = append(rsproto, sid...)
+	rsproto = append(rsproto, CR_LF...)
+
+	s.mu.Lock()
+	for _, r := range s.routes {
+		r.mu.Lock()
```

Comment:
Should check that we don't hold a route and want a server. We do pretty good at not walking back up the chain towards the server, but when I see these makes me think we need to double check for deadlock.

Reply:
When this function is invoked for RS+, it is in `checkForImportServices()` which is called in CLIENT, ROUTE and GATEWAY processing of inbound message without any lock, so that's ok.
For the RS-, this is called in `removeServiceImport()` which is invoked directly in `checkForImportServices()` (which we verified above is safe) and `pruneAutoExpireResponseMaps()` which is started as a go routine in some conditions from `addImplicitServiceImport()` (except in some tests when we invoke manually).
So as of now, this call is made without holding any specific lock.